File: go/linked-list/linked_list.go
// Package linkedlist provides a linked list.
package linkedlist

import (
	"errors"
)

// Node is an item in the List.
type Node struct {
	Value any
	prev  *Node
	next  *Node
}

// List is a linked list of Node.
type List struct {
	head *Node
	tail *Node
}

// ErrEmptyList is the error when an empty list is used.
var ErrEmptyList = errors.New("err empty list")

// Next returns the next element.
func (e *Node) Next() *Node {
	if e.next == nil || e.next.next == nil {
		return nil
	}
	return e.next
}

// Prev returns the previous element.
func (e *Node) Prev() *Node {
	if e.prev == nil || e.prev.prev == nil {
		return nil
	}
	return e.prev
}

func (e *Node) link() {
	e.next.prev = e
	e.prev.next = e
}

func (e *Node) unlink() any {
	e.next.prev, e.prev.next = e.prev, e.next
	return e.Value
}

// NewList returns a new list, populated with the values provided.
func NewList(args ...any) *List {
	l := &List{&Node{}, &Node{}}
	l.head.next = l.tail
	l.tail.prev = l.head
	for _, e := range args {
		l.Push(e)
	}
	return l
}

// PushFront pushes an element to the front of the list.
func (l *List) Unshift(v any) {
	(&Node{Value: v, prev: l.head, next: l.head.next}).link()
}

// Push pushes an element to the back of the list.
func (l *List) Push(v any) {
	(&Node{Value: v, prev: l.tail.prev, next: l.tail}).link()
}

// PopFront pops an element from the front of the list.
func (l *List) Shift() (any, error) {
	if l.isEmpty() {
		return nil, ErrEmptyList
	}
	return l.First().unlink(), nil
}

// PopBack pops an element from the back of the list.
func (l *List) Pop() (any, error) {
	if l.isEmpty() {
		return nil, ErrEmptyList
	}
	return l.Last().unlink(), nil
}

// Reverse returns a new list, reversed.
func (l *List) Reverse() *List {
	for e := l.head; e != nil; e = e.prev {
		e.prev, e.next = e.next, e.prev
	}
	l.head, l.tail = l.tail, l.head
	return l
}

// First returns the first element of a list.
func (l *List) First() *Node {
	if l.isEmpty() {
		return nil
	}
	return l.head.next
}

// Last returns the last element of a list.
func (l *List) Last() *Node {
	if l.isEmpty() {
		return nil
	}
	return l.tail.prev
}

func (l *List) isEmpty() bool {
	return l.head.next == l.tail
}

File: python/linked-list/linked_list.py
"""Doubly linked list."""

from __future__ import annotations
from typing import Callable, Generic, Generator, Optional, TypeVar
NodeValue = TypeVar('NodeValue')


class Node(Generic[NodeValue]):
    """List node."""
    def __init__(self, value: NodeValue):
        """Create a new node."""
        self.value = value
        self.prev: Optional[Node] = None
        self.next: Optional[Node] = None

    def add_prev(self, value: NodeValue) -> Node:
        """Add a node prior to self."""
        new_node = Node(value)
        assert self.prev is not None
        self.link(self.prev, new_node, self)
        return new_node

    def add_next(self, value: NodeValue) -> Node:
        """Add a node after self."""
        new_node = Node(value)
        assert self.next is not None
        self.link(self, new_node, self.next)
        return new_node

    @staticmethod
    def link(node_a: Node, node_b: Node, node_c: Node) -> None:
        """Set up node linkage between three nodes in a sequence."""
        node_a.next, node_b.next = node_b, node_c
        node_b.prev, node_c.prev = node_a, node_b

    def remove(self) -> NodeValue:
        """Remove a node from the list and return its value."""
        assert self.prev is not None
        assert self.next is not None
        self.next.prev = self.prev
        self.prev.next = self.next
        return self.value


def increment(
    func: Callable[[LinkedList, NodeValue], None]
) -> Callable[[LinkedList, NodeValue], None]:
    """LinkedList increment decorator to bump the length."""

    def wrapper(linked_list: LinkedList, value: NodeValue) -> None:
        func(linked_list, value)
        linked_list.length += 1

    return wrapper


def decrement(func: Callable[[LinkedList], NodeValue]) -> Callable[[LinkedList], NodeValue]:
    """LinkedList decrement decorator to check and reduce the length."""

    def wrapper(linked_list: LinkedList):
        if linked_list.length <= 0:
            raise IndexError("Cannot remove from an empty list.")
        linked_list.length -= 1
        return func(linked_list)

    return wrapper


class LinkedList:
    """A linked list."""

    def __init__(self):
        """Set up a linked list with an empty head and tail node."""
        # Create the head and last nodes.
        self.head = Node(None)
        self.last = Node(None)
        # Link the end nodes.
        self.head.next, self.last.prev = self.last, self.head
        self.length = 0

    @increment
    def push(self, value: NodeValue) -> None:
        """Push a node to the end of the list."""
        self.last.add_prev(value)

    @increment
    def unshift(self, value: NodeValue) -> None:
        """Insert a node at the start of the list."""
        self.head.add_next(value)

    @decrement
    def pop(self) -> NodeValue:
        """Pop a node from the end of the list."""
        return self.last.prev.remove()

    @decrement
    def shift(self) -> NodeValue:
        """Remove a node from the start of the list."""
        return self.head.next.remove()

    def __iter__(self) -> Generator[NodeValue, None, None]:
        """Iterate through the list."""
        cur = self.head.next
        while cur != self.last:
            yield cur.value
            cur = cur.next

    def __str__(self) -> str:
        """Return a string form."""
        return " -> ".join(repr(i) for i in self)

    def __len__(self) -> int:
        """Return the list length."""
        return self.length
